src/support.rs

```diff
@@ -20,48 +18,64 @@ pub trait FromRawSurface {
     unsafe fn from_raw_surface(surface: *mut ffi::cairo_surface_t) -> Self;
 }
 
+#[derive(Debug)]
+struct CallbackEnv<W> {
+    writer: W,
+    error: Option<io::Error>,
+}
+
 #[derive(Debug)]
 pub struct Writer<S: FromRawSurface + AsRef<Surface>, W: io::Write> {
     pub surface: S,
-    writer: Box<W>,
+    callback_env: Box<CallbackEnv<W>>,
 }
 
 impl<S: FromRawSurface + AsRef<Surface>, W: io::Write> Writer<S, W> {
-    extern fn write_cb(writer: *mut c_void, data: *mut c_uchar, length: c_uint) -> cairo_status_t {
-        let mut writer: Box<W> = unsafe { Box::from_raw(writer as *mut _) };
+    extern fn write_cb(env: *mut c_void, data: *mut c_uchar, length: c_uint) -> cairo_status_t {
+        let env: &mut CallbackEnv<W> = unsafe { &mut *(env as *mut CallbackEnv<W>) };
         let data = unsafe { slice::from_raw_parts(data, length as usize) };
 
-        let result = match writer.write_all(data) {
+        let result = match env.writer.write_all(data) {
             Ok(_) => Status::Success,
-            Err(_) => Status::WriteError,
+            Err(e) => {
+                env.error = Some(e);
+                Status::WriteError
+            }
         };
 
-        mem::forget(writer);
         result.into()
     }
 
     pub fn new(constructor: Constructor, width: f64, height: f64, writer: W) -> Writer<S, W> {
-        let mut writer = Box::new(writer);
-        let writer_ptr = unsafe { mem::transmute(&mut *writer) };
+        let mut callback_env = Box::new(CallbackEnv {
+            writer,
+            error: None,
+        });
+        let env_ptr = &mut *callback_env as *mut CallbackEnv<W> as *mut c_void;
         let surface = unsafe {
-            S::from_raw_surface(constructor(Some(Self::write_cb), writer_ptr, width, height))
+            S::from_raw_surface(constructor(Some(Self::write_cb), env_ptr, width, height))
         };
 
         Writer {
             surface,
-            writer,
+            callback_env,
         }
     }
 
-    pub fn writer(&self) -> &W { self.writer.as_ref() }
-    pub fn writer_mut(&mut self) -> &mut W { self.writer.as_mut() }
+    pub fn writer(&self) -> &W { &self.callback_env.writer }
+    pub fn writer_mut(&mut self) -> &mut W { &mut self.callback_env.writer }
 
-    pub fn finish(self) -> W {
-        let surface = self.surface;
-        surface.as_ref().finish();
-        drop(surface);
+    pub fn io_error(&self) -> Option<&io::Error> { self.callback_env.error.as_ref() }
+    pub fn take_io_error(&mut self) -> Option<io::Error> { self.callback_env.error.take() }
 
-        *self.writer
+    pub fn into_inner(self) -> Result<W, (io::Error, W)> {
```

Comment:
It would be nice if we could somehow force the user to call finish/into_inner at the end instead of letting the type be dropped silently. It's most likely a bug if it's not done, or am I missing something?

Not sure how we could do that nicely though.

Reply:
The type theory term for this is linear types (use exactly once). In Rust today, `!Copy` types are called affine (use at most once, where borrows do not count as a “use”). https://gankro.github.io/blah/linear-rust/ explains how it would be tricky to add linear types to the language.

Reply:
I was thinking of putting a panic into the `Drop` impl (ugly, at runtime) or somehow doing something with `warn_unused` but not sure how that could be nicely done here